File: netlib/src/main/java/net/ljb/kt/HttpConfig.java
package net.ljb.kt;

import net.ljb.kt.client.HttpClient;

import java.util.HashMap;
import java.util.Map;

public class HttpConfig {

    public interface ICommHeader {
        void addCommHeader(Map<String, String> headers);
    }

    public interface ICommParam {
        void addCommHeader(Map<String, String> params);
    }

    public interface ICommCookie {
        void saveCookie(String host, String cookie);

        String loadCookie(String host);
    }

    private final String baseUrl;
    private final boolean openLog;
    private final ICommHeader commHeader;
    private final ICommParam commParam;
    private final ICommCookie commCookie;


    private HttpConfig(String baseUrl, ICommHeader header, ICommParam param, ICommCookie cookie, boolean openLog) {
        this.baseUrl = baseUrl;
        this.commHeader = header;
        this.commParam = param;
        this.commCookie = cookie;
        this.openLog = openLog;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    public boolean isOpenLog() {
        return openLog;
    }

    public Map<String, String> getCommParam() {
        if (commParam == null) return null;
        HashMap<String, String> map = new HashMap<>();
        commParam.addCommHeader(map);
        return map;
    }

    public Map<String, String> getCommHeader() {
        if (commHeader == null) return null;
        HashMap<String, String> map = new HashMap<>();
        commHeader.addCommHeader(map);
        return map;
    }


    public ICommCookie getCommCookie() {
        return commCookie;
    }

    public static class Builder {

        private final String baseUrl;
        private boolean openLog;
        private ICommHeader commHeader;
        private ICommParam commParam;
        private ICommCookie commCookie;

        public Builder(String baseUrl) {
            this.baseUrl = baseUrl;
        }

        public Builder addCommCookie(ICommCookie cookie) {
            this.commCookie = cookie;
            return this;
        }

        public Builder addCommHeader(ICommHeader header) {
            this.commHeader = header;
            return this;
        }

        public Builder addCommParam(ICommParam param) {
            this.commParam = param;
            return this;
        }

        public Builder openLog(boolean open) {
            this.openLog = open;
            return this;
        }

        public void build() {
            HttpConfig httpConfig = new HttpConfig(baseUrl, commHeader, commParam, commCookie, openLog);
            HttpClient.INSTANCE.init(httpConfig);
        }

    }

}

File: netlib/src/main/java/net/ljb/kt/HttpConfig.kt
package net.ljb.kt

import net.ljb.kt.client.HttpClient
import okhttp3.OkHttpClient
import java.util.*

class HttpConfig private constructor(
    val baseUrl: String,
    private val commHeader: ((headers: Map<String, String>) -> Unit)?,
    private val commParam: ((params: Map<String, String>) -> Unit)?,
    val commCookie: ICommCookie?,
    val isOpenLog: Boolean,
    var tag: String
) {

    var client: OkHttpClient? = null

    interface ICommCookie {
        fun saveCookie(host: String, cookie: String)
        fun loadCookie(host: String): String?
    }

    fun getCommParam(): Map<String, String>? {
        if (commParam == null) return null
        val map = HashMap<String, String>()
        commParam.invoke(map)
        return map
    }

    fun getCommHeader(): Map<String, String>? {
        if (commHeader == null) return null
        val map = HashMap<String, String>()
        commHeader.invoke(map)
        return map
    }

    class Builder(private val baseUrl: String) {

        private var openLog = false
        private var commHeader: ((headers: Map<String, String>) -> Unit)? = null
        private var commParam: ((params: Map<String, String>) -> Unit)? = null
        private var commCookie: ICommCookie? = null
        private var tag: String = HttpClient.TAG_DEFAULT_CLIENT

        fun addCommCookie(cookie: ICommCookie?): Builder {
            commCookie = cookie
            return this
        }

        fun addCommHeader(header: ((params: Map<String, String>) -> Unit)): Builder {
            commHeader = header
            return this
        }

        fun addCommParam(param: ((params: Map<String, String>) -> Unit)): Builder {
            commParam = param
            return this
        }

        fun openLog(open: Boolean): Builder {
            openLog = open
            return this
        }

        fun newClientByTag(tag: String): Builder {
            this.tag = tag
            return this
        }

        fun build() {
            HttpClient.init(
                HttpConfig(
                    baseUrl,
                    commHeader,
                    commParam,
                    commCookie,
                    openLog,
                    tag
                )
            )
        }
    }
}